marlowe-symbolic/src/Language/Marlowe/Analysis/FSSemantics.hs

```diff
@@ -1,1126 +1,415 @@
+     return (ite cond newTrace sFalse)
+isValidAndFailsWhen (Case (Deposit accId party token val) cont:rest)
+                    timeout timCont previousMatch sState pos =
+  do let concVal = symEvalVal val sState
+     let symInput = SymDeposit accId party token concVal
+     let clashResult = previousMatch symInput
+     let newPreviousMatch otherSymInput =
+           case otherSymInput of
+             SymDeposit otherAccId otherParty otherToken otherConcVal ->
+               if (otherAccId == accId) && (otherParty == party)
+                  && (otherToken == token)
+               then (otherConcVal .== concVal) .|| previousMatch otherSymInput
+               else previousMatch otherSymInput
+             _ -> previousMatch otherSymInput
+     ((newCond, newSState), newTrace)
+               <- applyInputConditions (Just symInput) timeout sState pos cont
+     contTrace <- isValidAndFailsWhen rest timeout timCont newPreviousMatch sState (pos + 1)
+     return (ite (newCond .&& newTrace)
+                 newTrace
+                 ((newCond .&& (concVal .<= 0) -- Non-positive deposit warning
+                           .&& convertToSymbolicTrace (traces newSState)
+                                                      (paramTrace sState))
+                  .|| contTrace))
+isValidAndFailsWhen (Case (Choice choId bnds) cont:rest)
+                    timeout timCont previousMatch sState pos =
+  do concVal <- generateValueInBounds bnds
+     let symInput = SymChoice choId concVal
+     let clashResult = previousMatch symInput
+     let newPreviousMatch otherSymInput =
+           case otherSymInput of
+             SymChoice otherChoId otherConcVal ->
+               if otherChoId == choId
+               then (otherConcVal .== concVal) .|| previousMatch otherSymInput
+               else previousMatch otherSymInput
+             _ -> previousMatch otherSymInput
+     contTrace <- isValidAndFailsWhen rest timeout timCont newPreviousMatch sState (pos + 1)
+     ((newCond, newSState), newTrace)
+               <- applyInputConditions (Just symInput) timeout sState pos cont
+     return (ite (newCond .&& newTrace) newTrace contTrace)
+isValidAndFailsWhen (Case (Notify obs) cont:rest)
+                    timeout timCont previousMatch sState pos =
+  do let obsRes = symEvalObs obs sState
+     let symInput = SymNotify
+     let clashResult = previousMatch symInput
+     let newPreviousMatch otherSymInput =
+           case otherSymInput of
+             SymNotify -> sNot obsRes
+             _         -> previousMatch otherSymInput
+     (newCond, newSState) <- addTransaction (Just symInput) timeout sState pos
+     newTrace <- isValidAndFailsAux cont newSState
+     contTrace <- isValidAndFailsWhen rest timeout timCont newPreviousMatch sState (pos + 1)
+     return (ite (newCond .&& obsRes .&& newTrace) newTrace contTrace)
+
+-- SBV handling and result extraction
+
+countWhens :: Contract -> Integer
```

Comment:
`contractDepth` maybe?

Reply:
No, that is something else, maybe countMaxNestedWhens, but well, this is copied from Isabelle semantics.